client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +28,388 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+          return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot not ready");
+          break;
+        case SnapshotResponseStatus::Success:
+          if ((res->data).snapshot_id != snapshot_id) {
+            return_status = grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, "Hash mismatch from replicas");
+          }
+          if ((res->data).hash != final_hash) {
+            return_status = grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, "Hash mismatch from replicas");
+          }
+          break;
+        default:
+          ConcordAssert(false);
+          break;
+      }
+    } else {
+      break;  // Break the for loop abruptly
+    }
+  }
 }
 
 Status StateSnapshotServiceImpl::ReadAsOf(ServerContext* context,
                                           const ReadAsOfRequest* proto_request,
                                           ReadAsOfResponse* response) {
-  // TODO: Add implementation
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "ReadAsOf");
+  std::chrono::milliseconds timeout = 5s;  // This is default timeout
+  if (context != nullptr) {
+    auto time_left =
+        gpr_time_to_millis(gpr_time_sub(gpr_now(context->raw_deadline().clock_type), context->raw_deadline()));
+    std::chrono::milliseconds curr_timeout{time_left};
+    if (curr_timeout.count() > 0) {
+      timeout = curr_timeout;
+    }
+  }
+
+  ReadConfig read_config{RequestConfig{false, 0, 1024 * 1024, timeout, "readasofreq", "", false, true},
+                         bft::client::ByzantineSafeQuorum{}};
+
+  std::vector<std::unique_ptr<StateSnapshotReadAsOfResponse>> read_as_of_responses;
+  grpc::Status return_status;
+  auto callback =
+      [&read_as_of_responses, &return_status, &read_config](concord::client::concordclient::SendResult&& send_result) {
+        ResponseType<StateSnapshotReadAsOfResponse> reponse;
+        getResponseSetStatus(std::move(send_result),
+                             reponse,
+                             return_status,
+                             read_config.request.correlation_id,
+                             true,
+                             "concord.client.clientservice.state_snapshot_service.signedhash.callback");
+        read_as_of_responses =
+            std::get<std::vector<std::unique_ptr<StateSnapshotReadAsOfResponse>>>(std::move(reponse.response));
+      };
+
+  ConcordAssertNE(proto_request, nullptr);
+  ReconfigurationRequest rreq;
+  StateSnapshotReadAsOfRequest cmd;
+  cmd.snapshot_id = proto_request->snapshot_id();
+  cmd.participant_id = client_->getSubscriptionId();
+  for (int i = 0; i < proto_request->keys_size(); ++i) {
+    cmd.keys.push_back(proto_request->keys(i));
+  }
+  rreq.command = cmd;
+  bft::client::Msg message;
+  concord::messages::serialize(message, rreq);
+  client_->send(read_config, std::move(message), callback);
+  for (const auto& res : read_as_of_responses) {
+    if (return_status.ok()) {
```

Comment:
The following code gets nested into 8 levels!! This is hard to keep track of when reading the code. Given that we `break` if the return status is not ok, I recommend to do this right here to get rid of one level of nesting.